### PR TITLE
Fast-forward Langevin functionality included

### PR DESCRIPTION
**Pull request by dilkins**
## Purpose

Implementation of the Fast-Forward Langevin (FFL) thermostat, as described in J. Chem. Phys. 148, 184109 (2018).

## Author(s)

Lionel Constantin, David M. Wilkins, Michele Ceriotti, EPFL Lausanne

## Backward Compatibility

N/A

## Implementation Notes

The code for FFL dynamics was based on fix_gle, the code for Generalized Langevin Equation dynamics. Correctness of the algorithm was verified by comparison with the results in Fig. 4 of the [paper](https://aip.scitation.org/doi/full/10.1063/1.5029833). No other LAMMPS features are affected.

## Post Submission Checklist

_Please check the fields below as they are completed_
- [ ] The feature or features in this pull request is complete
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] One or more example input decks are included
- [ ] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

[Link to Fast-Forward Langevin paper](https://aip.scitation.org/doi/full/10.1063/1.5029833)

### CODE REVIEW

**Comment by akohlmey** [comment]
@sjplimp since we were discussing renaming of styles in the other pull request, seeing this PR reminds me, that we now have multiple langevin thermostat variants (and with this one more), and perhaps we should consider renaming them to langevin/XXX to make their heritage more visible. OTOH, the naming style here (initials of the name) is consistent with several styles (not just langevin thermostats). Do you have a preference?

File: src/USER-MISC/fix_ffl.cpp [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+   M. Hijazi, D. M. Wilkins, M. Ceriotti, J. Chem. Phys. 148, 184109 (2018)
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing authors: Lionel Constantin (EPFL), David M. Wilkins (EPFL), 
+			 Michele Ceriotti (EPFL)
+------------------------------------------------------------------------- */
+
+#include <mpi.h>
+#include <math.h>
```
Comment (akohlmey) [inline src/USER-MISC/fix_ffl.cpp]: please use c++ style system headers here: <cmath> instead of <math.h>, <cstring> instead of <string.h> and <cstdlib> instead of <stdlib.h>.

File: src/USER-MISC/fix_ffl.cpp [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+//#define FFL_DEBUG 1
+
+#define MAXLINE 1024
+
+/* syntax for fix_ffl:
+ * fix nfix id-group ffl tau Tstart Tstop seed [flip_type]
+ *
+ *                                                                        */
+
+/* ---------------------------------------------------------------------- */
```
Comment (akohlmey) [inline src/USER-MISC/fix_ffl.cpp]: For improved consistency with other sources in LAMMPS, we would appreciate, if this could be reformatted to follow the LAMMPS coding style a bit closer. The main differences i currently see is, that usually we use 2 character indentation and put the opening brace of if/for/while statements on the same line, not the next. The main benefit is that diffs become more readable, in case we need to make changes in the future and then the automatic re-indentation would hide the real changes. This is a hard requirement only for files in standard packages, which this is not. So you don't need to do it, if it is too much of a bother.

File: src/USER-MISC/fix_ffl.cpp [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+
+    // final temperature (t ramp)
+    t_stop = force->numeric(FLERR,arg[5]);
+
+    // PRNG seed
+    int seed = force->inumeric(FLERR,arg[6]);
+
+    // Flip type used, uses rescale if no flip is given
+    if (narg == 8) strcpy(flip_type,arg[7]);
+    else strcpy(flip_type,"rescale");
```
Comment (akohlmey) [inline src/USER-MISC/fix_ffl.cpp]: Please use strncpy() here or make flip_type dynamically allocated, so that flip_type does not overflow and cause a crash on incorrect input.

Comment (akohlmey) [reply src/USER-MISC/fix_ffl.cpp]: ... or use an enumerator as it is done in many similar cases in LAMMPS.

**Comment by ceriottm** [comment]
@athomps  I agree it might be nice to "merge" with GLE. main reasons why we did not, and I'm reluctant to do that unless it's a showstopper, are that (1) we haven't really tested how this idea would generalize to the GLE framework (i.e. do you "flip" also the additional momenta or not?) and (2) this simple version should be easily combined with rigid-body integrators, whereas the GLE with rigid bodies is a can of worms that we tried to open a couple of times but ran away in terror.

File: src/USER-MISC/fix_ffl.cpp (hunk 2 above)
Comment (ceriottm) [reply src/USER-MISC/fix_ffl.cpp]: Will do it, makes sense since it's a small package anyway

**Comment by akohlmey** [comment]
@ceriottm @athomps i don't consider merging/consolidating langevin thermostats a showstopper issue. if we would want to do that, i would expand the scope and do this in a more general way, since there are more langevin based thermostats, and they come in different flavors (e.g. with/without time integration). from what i know, there may be more langevin features added in the near future, and i believe it would be better to think about consolidation after that. also, i would have to say, that thermostats in general have been causing very low maintenance effort, if they are well tested at submission time. so the penalty for not merging is not so large as it might seem.

in short, i would suggest we move forward with doing the few small formal adjustments request and then merge this feature afterwards, and then perhaps file a TODO item in the issue tracker to remind us, to make an effort to generalize langevin thermostats across the board (at which point, we can possibly also rethink the naming conventions for those).

**Comment by ceriottm** [comment]
@akohlmey OK so I'll push @dilkins and Lionel to go through the style changes quickly. 
actually I'm at a conference with Ben Leimkuhler and I'll discuss with him about rigid bodies with GLE, we might try and implement that in LAMMPS, and that would be a good opportunity for consolidation

**Comment by athomps** [comment]
Thanks for the explanation. I am okay with keeping GLE and FFL separate.

**Comment by dilkins** [comment]
Thanks all for your comments: I've taken them on board in the two new commits.

**Comment by dilkins** [comment]
Apologies, yes, I've included this in d5e71e709913271d5d192c0904c216e2741b3323
